File: nodejs-common/CardPaymentRequest.js
/*
 *  Copyright 2006-2016 WebPKI.org (http://webpki.org).
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

'use strict';

// Saturn "CardPaymentRequest" object

const JsonUtil = require('webpki.org').JsonUtil;
const ByteArray = require('webpki.org').ByteArray;

const BaseProperties        = require('./BaseProperties');
const Messages              = require('./Messages');
const Software              = require('./Software');
const AuthorizationResponse = require('./AuthorizationResponse');
const AuthorizationRequest  = require('./AuthorizationRequest');
const ProtectedAccountData  = require('./ProtectedAccountData');

function CardPaymentRequest(rd) {
  this.root = Messages.parseBaseMessage(Messages.CARD_PAYMENT_REQUEST, rd);
  this.authorizationResponse = new AuthorizationResponse(rd.getObject(BaseProperties.EMBEDDED_JSON));
  this.referenceId = rd.getString(BaseProperties.REFERENCE_ID_JSON);
  this.timeStamp = rd.getDateTime(BaseProperties.TIME_STAMP_JSON);
  this.software = new Software(rd);
  this.publicKey = rd.getSignature().getPublicKey();
  AuthorizationRequest.comparePublicKeys(this.publicKey,
                                         this.authorizationResponse.authorizationRequest.paymentRequest);
  if (!this.authorizationResponse.authorizationRequest.payerAccountType.isCardPayment()) {
    throw new TypeError('Payment method is not card: ' + 
        this.authorizationResponse.authorizationRequest.payerAccountType.getTypeUri());
  }
  rd.checkForUnread();
}

CardPaymentRequest.prototype.getTimeStamp = function() {
  return this.timeStamp;
};

CardPaymentRequest.prototype.getReferenceId = function() {
  return this.referenceId;
};

CardPaymentRequest.prototype.getPayee = function() {
  return this.authorizationResponse.authorizationRequest.paymentRequest.payee;
};

CardPaymentRequest.prototype.getPublicKey = function() {
  return this.publicKey;
};

CardPaymentRequest.prototype.getAuthorizationResponse = function() {
  return this.authorizationResponse;
};

CardPaymentRequest.prototype.getPaymentRequest = function() {
  return this.authorizationResponse.authorizationRequest.paymentRequest;
};

CardPaymentRequest.prototype.verifyPayerProvider = function(paymentRoot) {
  this.authorizationResponse.signatureDecoder.verifyTrust(paymentRoot);
};

CardPaymentRequest.prototype.getProtectedAccountData = function(decryptionKeys) {
  return new ProtectedAccountData(new JsonUtil.ObjectReader(JSON.parse(ByteArray.utf8ToString(
                                    this.authorizationResponse
                                      .encryptedAccountData
                                        .getDecryptedData(decryptionKeys)))),
                                  this.authorizationResponse.authorizationRequest.payerAccountType);
};

/*
    public static JSONObjectWriter encode(AuthorizationResponse authorizationResponse,
                                          String referenceId,
                                          ServerAsymKeySigner signer) throws IOException {
        return Messages.createBaseMessage(Messages.CARD_PAYMENT_REQUEST)
            .setObject(EMBEDDED_JSON, authorizationResponse.root)
            .setString(REFERENCE_ID_JSON, referenceId)
            .setDateTime(TIME_STAMP_JSON, new Date(), true)
            .setObject(SOFTWARE_JSON, Software.encode(AuthorizationRequest.SOFTWARE_NAME, 
                                                      AuthorizationRequest.SOFTWARE_VERSION))
            .setSignature(signer);
    }

    public ProtectedAccountData getProtectedAccountData(Vector<DecryptionKeyHolder> decryptionKeys, boolean cardAccount)
    throws IOException, GeneralSecurityException {
        return new ProtectedAccountData(JSONParser.parse(authorizationResponse
                                                             .encryptedAccountData
                                                                 .getDecryptedData(decryptionKeys)), cardAccount);
    }

    public void verifyUserBank(JSONX509Verifier paymentRoot) throws IOException {
        authorizationResponse.signatureDecoder.verify(paymentRoot);
    }
*/

module.exports = CardPaymentRequest;

File: resources/common/org/webpki/saturn/common/CardPaymentRequest.java
/*
 *  Copyright 2006-2016 WebPKI.org (http://webpki.org).
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
package org.webpki.saturn.common;

import java.io.IOException;

import java.math.BigDecimal;

import java.security.GeneralSecurityException;
import java.security.PublicKey;

import java.util.GregorianCalendar;
import java.util.Vector;

import org.webpki.crypto.AlgorithmPreferences;

import org.webpki.json.JSONObjectReader;
import org.webpki.json.JSONObjectWriter;
import org.webpki.json.JSONParser;
import org.webpki.json.JSONX509Verifier;

import org.webpki.json.encryption.DecryptionKeyHolder;

public class CardPaymentRequest implements BaseProperties {
    
    public CardPaymentRequest(JSONObjectReader rd, Boolean cardNetwork) throws IOException {
        Messages.parseBaseMessage(Messages.CARD_PAYMENT_REQUEST, root = rd);
        authorizationResponse = new AuthorizationResponse(rd.getObject(EMBEDDED_JSON));
        actualAmount = rd.getBigDecimal(AMOUNT_JSON,
                                        authorizationResponse
                                            .authorizationRequest.paymentRequest.currency.decimals);
        referenceId = rd.getString(REFERENCE_ID_JSON);
        timeStamp = rd.getDateTime(TIME_STAMP_JSON);
        software = new Software(rd);
        publicKey = rd.getSignature(AlgorithmPreferences.JOSE).getPublicKey();
        AuthorizationRequest.comparePublicKeys(publicKey,
                                               authorizationResponse
                                                   .authorizationRequest.paymentRequest);
        if (cardNetwork != null &&
            authorizationResponse.authorizationRequest.payerAccountType.isCardPayment() ^ cardNetwork) {
            throw new IOException("Incompatible payment method: " + 
                authorizationResponse.authorizationRequest.payerAccountType.getTypeUri());
        }
        rd.checkForUnread();
    }

    Software software;

    JSONObjectReader root;

    GregorianCalendar timeStamp;
    public GregorianCalendar getTimeStamp() {
        return timeStamp;
    }

    BigDecimal actualAmount;
    public BigDecimal getAmount() {
        return actualAmount;
    }

    String referenceId;
    public String getReferenceId() {
        return referenceId;
    }

    public boolean getTestMode() {
        return authorizationResponse.authorizationRequest.testMode;
    }

    public Payee getPayee() {
        return authorizationResponse.authorizationRequest.paymentRequest.payee;
    }

    public PaymentRequest getPaymentRequest() {
        return authorizationResponse.authorizationRequest.paymentRequest;
    }

    PublicKey publicKey;
    public PublicKey getPublicKey() {
        return publicKey;
    }

    AuthorizationResponse authorizationResponse;
    public AuthorizationResponse getAuthorizationResponse() {
        return authorizationResponse;
    }

    public static JSONObjectWriter encode(AuthorizationResponse authorizationResponse,
                                          BigDecimal actualAmount,
                                          String referenceId,
                                          ServerAsymKeySigner signer) throws IOException {
        return Messages.createBaseMessage(Messages.CARD_PAYMENT_REQUEST)
            .setObject(EMBEDDED_JSON, authorizationResponse.root)
            .setBigDecimal(AMOUNT_JSON,
                           actualAmount,
                           authorizationResponse.authorizationRequest.paymentRequest.currency.decimals)
            .setString(REFERENCE_ID_JSON, referenceId)
            .setDateTime(TIME_STAMP_JSON, new GregorianCalendar(), true)
            .setObject(SOFTWARE_JSON, Software.encode(AuthorizationRequest.SOFTWARE_NAME, 
                                                      AuthorizationRequest.SOFTWARE_VERSION))
            .setSignature(signer);
    }

    public ProtectedAccountData getProtectedAccountData(Vector<DecryptionKeyHolder> decryptionKeys)
    throws IOException, GeneralSecurityException {
        return new ProtectedAccountData(JSONParser.parse(authorizationResponse
                                                             .encryptedAccountData
                                                                 .getDecryptedData(decryptionKeys)),
                                        authorizationResponse.authorizationRequest.payerAccountType);
    }

    public void verifyUserBank(JSONX509Verifier paymentRoot) throws IOException {
        authorizationResponse.signatureDecoder.verify(paymentRoot);
    }
}
